### ISSUE TITLE
great stuff - how about turning it into a ImageIO CoDec?

### ISSUE THREAD

**Issue by elbosso** [Traneptora/jxlatte#20]
I would be willing to participate in the effort as i wrote a few ImageIO codecs already - though not as complex ones as would be needed here... 

One could maybe get inspiration for example from here: https://xmlgraphics.apache.org/batik/javadoc/org/apache/batik/ext/awt/image/codec/imageio/ImageIOJPEGImageWriter.html

**Comment by Traneptora** [comment]
It's on the list of things to do.

Also the file you linked is a writer, not a reader.

**Comment by elbosso** [comment]
yep, I know. But maybe this could be a far away stretch goal. I assume, that one would also find a reader somewhere in there...

**Comment by elbosso** [comment]
ok - a crude example implementation is done as a prototype. I can use my branch to load JPEG XL using ImageIO - however, there remains the issue that the BufferedImages created by this library seem to be incompatible with ImageIO (compare #22)

**Comment by haraldk** [comment]
Hi,

I'm considering implementing an ImageIO plugin (`ImageReader` only for now) based on this project. Either by forking, or using this project as a library dependency.

I see issue #22 refers to a method `JXImage.asBufferedImage` or `toBufferedImage`, however I don't find these in the current code base..? Was this method removed, or was it only available in a special branch?

**Comment by Traneptora** [comment]
I removed all the BufferedImage and WritableRaster code because `java.awt` isn't available on android. Currently `JXLImage` produces an array of `ImageBuffer` which can be then turned into a `BufferedImage` by using a `WritableRaster` and a `BandedSampleModel`.

**Comment by haraldk** [comment]
@Traneptora Thanks! After looking some more at the code and testing with various sample images, I can only way, this is seriously impressive stuff! 

I completely understand why you want to keep any AWT dependencies out of the project, keeping it like this makes things more flexible. 👍🏻 

I've made an adapter class that can convert to `BufferedImage`s in a more "standard" types (8 or 16 bit pixel interleaved), depending on the bit count in the header. This should allow me to create an `ImageIO` plugin that creates images that can be easily displayed or converted to other formats. 

I have loads of questions though, but I might open another issue for that. 😀

**Comment by Traneptora** [comment]
It's recommended to not use `ImageIO.write()` to output PNG files because it notably doesn't write any color information. Also be aware that some JXL files are HDR and thus jxlatte decodes them to an HDR image. If you need images in a particular color space you may want to call `JXLImage.transform` first.

**Comment by elbosso** [comment]
> Hi,
> 
> I'm considering implementing an ImageIO plugin (`ImageReader` only for now) based on this project. Either by forking, or using this project as a library dependency.
> 
> I see issue #22 refers to a method `JXImage.asBufferedImage` or `toBufferedImage`, however I don't find these in the current code base..? Was this method removed, or was it only available in a special branch?
> 
> Update: I found the code in @elbosso's branch, and was able to adapt it to the current main branch + convert the images to more compatible `BufferedImage` types. I'm not super-familiar with JPEG XL, so I might be missing why the library uses 32 bit float or integer components for what is normally 8-16 bits...?

I am glad I could be of some (little) help and looking forward to the final result!

**Comment by haraldk** [comment]
> It's recommended to not use `ImageIO.write()` to output PNG files because it notably doesn't write any color information to the PNG file (this is why I rolled my own primitive png writer). 

I'm not sure if I understand what you mean? Do you mean that it would not write PNG files from the `toBufferedImage` method in the old branch? In that case, I have fixed this issue. 
Or do you mean it does not correctly preserve color information, like ICC profiles or HDR?

> If you need images in a particular color space you may want to call `JXLImage.transform` first.

Excellent, that is actually what I needed! I had some issue with the images coming out way too "warn". Now, after doing the same transform as you do in your PNGWriter the images look just the same (I haven't compared pixel values directly, but they look equal).

samples/bbb.jxl transcoded to PNG using JXLatte `PNGWriter`:

![bbb](https://github.com/user-attachments/assets/dd1702cd-0a7b-4849-b72e-14ba2e7945ee)

samples/bbb.jxl transcoded to PNG using my preliminary code and `ImageIO.write` (ie. the built in `PNGImageWriter`):

![bbb jxl_tm](https://github.com/user-attachments/assets/0844ddec-3887-4637-bff6-133bd026a0f4)

**Comment by Traneptora** [comment]
Currently there is not. It is something I could do, but I haven't added it yet.
